File: src/part3/_Lim2156.java
package part3;

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class _Lim2156 {	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		int[] wines = new int[n];
		int[] maxes = new int[n];
		
		for(int i = 0; i < n; i++) {
			wines[i] = Integer.parseInt(br.readLine());
		}
		
		if(n >= 3) {
			maxes[0] = wines[0];
			maxes[1] = wines[0] + wines[1];
			maxes[2] = Math.max(wines[0] + wines[2], wines[1] + wines[2]);
			
			for(int i = 3; i < n; i++) {
				maxes[i] = Math.max(maxes[i - 2] + wines[i], maxes[i - 3] + wines[i - 1] + wines[i]);
			}
			
			System.out.println(Math.max(maxes[n - 1], maxes[n - 2]));
		} else {
			int sum = 0;
			
			for(int i = 0; i < n; i++) {
				sum += wines[i];
			}
			
			System.out.println(sum);
		} 
	}
}

File: src/main/java/boj/part7/Lim2156.kt
package boj.part7

import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val wines = IntArray(10_000)
    repeat(n) {
        wines[it] = br.readLine().toInt()
    }

    val maxes = IntArray(10_000)
    maxes[0] = wines[0]
    maxes[1] = wines[1] + wines[0]
    maxes[2] = maxOf(maxes[1], wines[2] + wines[0], wines[2] + wines[1])

    for (i in 3 until n) {
        maxes[i] = maxOf(maxes[i - 1], wines[i] + maxes[i - 2], wines[i] + wines[i - 1] + maxes[i - 3])
    }

    println(maxes[n - 1])
}
